### ISSUE TITLE
Is it possible to use an app on a virtual display while having the phone screen turned off?

### ISSUE THREAD

**Issue by lKomus** [Genymobile/scrcpy#5522]
Hello,

The new option for a virtual display seems to work great for me. Before that I was using to play some games on my computer while turning off the phone screen to save some battery with "-Sw" tags. With the virtual display I can have the game take the entire space of my screen, which is great.
Unfortunately, if I turn off my phone's screen, the virtual display is also locked, so I'm forced to have my phone screen constantly turned on while I'm using the virtual display.

Am I missing an option or is this an Android limitation?

**Comment by rom1v** [comment]
It probably depends on the device and Android version. On a Pixel 8, I can open a virtual display without unlocking the device.

As a workaround, can you turn the screen off with Alt+o (or --turn-screen-off)?

**Comment by lKomus** [comment]
My phone is a Poco F2 Pro on MIUI 14 with Android 12.

--turn-screen-off doesn't work, neither does Alt + o. Alt + p works but it locks the virtual display with the phone.

**Comment by rom1v** [comment]
> --turn-screen-off doesn't work, neither does Alt + o

OK, that's because they try to turn off the displayId you're mirroring, not the main display.

Could you try with this change:

```diff
diff --git server/src/main/java/com/genymobile/scrcpy/device/Device.java server/src/main/java/com/genymobile/scrcpy/device/Device.java
index cd7134993..18becab6e 100644
--- server/src/main/java/com/genymobile/scrcpy/device/Device.java
+++ server/src/main/java/com/genymobile/scrcpy/device/Device.java
@@ -127,6 +127,8 @@ public final class Device {
     public static boolean setDisplayPower(int displayId, boolean on) {
         assert displayId != Device.DISPLAY_ID_NONE;
 
+        displayId = 0;
+
         if (Build.VERSION.SDK_INT >= AndroidVersions.API_35_ANDROID_15) {
             return ServiceManager.getDisplayManager().requestDisplayPower(displayId, on);
         }
```

**Comment by lKomus** [comment]
Um, I'm sorry I don't think I know how to do this... Do I have to build the app with this change? I'll figure it out but I've never done this before.

**Comment by lKomus** [comment]
Ok, it works now, Alt + o works, -Sw works, and --turn-screen-off works. The only caveat is that when I close scrcpy, the screen is still turned off and I have to press power twice on my phone to force it back on.

Unrelated to the issue, each time I close and reopen scrcpy, the virtual display id goes up by 1 everytime, I think it went up to like 18 by now.

Thanks for the quick fix tho!

**Comment by rom1v** [comment]
> Unrelated to the issue, each time I close and reopen scrcpy, the virtual display id goes up by 1 everytime, I think it went up to like 18 by now.

Yes, that's expected.

> The only caveat is that when I close scrcpy, the screen is still turned off and I have to press power twice on my phone to force it back on.

Please test with:

```diff
diff --git server/src/main/java/com/genymobile/scrcpy/CleanUp.java server/src/main/java/com/genymobile/scrcpy/CleanUp.java
index 352f7c6b7..911cb5ba6 100644
--- server/src/main/java/com/genymobile/scrcpy/CleanUp.java
+++ server/src/main/java/com/genymobile/scrcpy/CleanUp.java
@@ -207,13 +207,15 @@ public final class CleanUp {
             }
         }
 
-        if (displayId != Device.DISPLAY_ID_NONE && Device.isScreenOn(displayId)) {
+        // Change the display power of the main display when mirroring a virtual display
+        int targetDisplayId = displayId != Device.DISPLAY_ID_NONE ? displayId : 0;
+        if (Device.isScreenOn(targetDisplayId)) {
             if (powerOffScreen) {
                 Ln.i("Power off screen");
-                Device.powerOffScreen(displayId);
+                Device.powerOffScreen(targetDisplayId);
             } else if (restoreDisplayPower) {
                 Ln.i("Restoring display power");
-                Device.setDisplayPower(displayId, true);
+                Device.setDisplayPower(targetDisplayId, true);
             }
         }
 
diff --git server/src/main/java/com/genymobile/scrcpy/control/Controller.java server/src/main/java/com/genymobile/scrcpy/control/Controller.java
index f0e4c0374..fb9eba105 100644
--- server/src/main/java/com/genymobile/scrcpy/control/Controller.java
+++ server/src/main/java/com/genymobile/scrcpy/control/Controller.java
@@ -281,7 +281,7 @@ public class Controller implements AsyncProcessor, VirtualDisplayListener {
                 setClipboard(msg.getText(), msg.getPaste(), msg.getSequence());
                 break;
             case ControlMessage.TYPE_SET_DISPLAY_POWER:
-                if (supportsInputEvents && displayId != Device.DISPLAY_ID_NONE) {
+                if (supportsInputEvents) {
                     setDisplayPower(msg.getOn());
                 }
                 break;
@@ -691,7 +691,9 @@ public class Controller implements AsyncProcessor, VirtualDisplayListener {
     }
 
     private void setDisplayPower(boolean on) {
-        boolean setDisplayPowerOk = Device.setDisplayPower(displayId, on);
+        // Change the display power of the main display when mirroring a virtual display
+        int targetDisplayId = displayId != Device.DISPLAY_ID_NONE ? displayId : 0;
+        boolean setDisplayPowerOk = Device.setDisplayPower(targetDisplayId, on);
         if (setDisplayPowerOk) {
             keepDisplayPowerOff = !on;
             Ln.i("Device display turned " + (on ? "on" : "off"));
```

- [`scrcpy-server`](https://tmp.rom1v.com/scrcpy/5522/2/scrcpy-server) <sub>`SHA-256: 39ea012cbc529cc93d8da9332d4c090ef916c1ce5c169a1737e370a5de5ed0b`</sub>

**Comment by lKomus** [comment]
That works! Thanks a lot!

**Comment by rom1v** [comment]
Released in 3.0.1.
